apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetItemService.java

```diff
@@ -207,6 +224,116 @@ private NotFoundException failWithNotFound(String message) {
+                })
+                .then();
+    }
+
+    /**
+     * Creates a new DatasetItem with enriched data containing trace metadata.
+     * 
+     * @param item the original dataset item
+     * @param trace the trace containing metadata to add
+     * @return a new DatasetItem with enriched data
+     */
+    private DatasetItem enrichItemWithTraceMetadata(DatasetItem item, Trace trace) {
+        // Create a mutable copy of the data map
+        Map<String, JsonNode> enrichedData = new HashMap<>(item.data());
+
+        // Add trace metadata to the data map
+        if (trace.tags() != null && !trace.tags().isEmpty()) {
+            enrichedData.put("trace_tags", JsonUtils.getJsonNodeFromStringWithFallback(
+                    JsonUtils.writeValueAsString(trace.tags())));
+        }
+
+        if (trace.comments() != null && !trace.comments().isEmpty()) {
+            enrichedData.put("trace_comments", JsonUtils.getJsonNodeFromStringWithFallback(
+                    JsonUtils.writeValueAsString(trace.comments())));
+        }
+
+        if (trace.feedbackScores() != null && !trace.feedbackScores().isEmpty()) {
+            enrichedData.put("trace_feedback_scores", JsonUtils.getJsonNodeFromStringWithFallback(
+                    JsonUtils.writeValueAsString(trace.feedbackScores())));
+        }
+
+        if (trace.guardrailsValidations() != null && !trace.guardrailsValidations().isEmpty()) {
+            enrichedData.put("trace_guardrails_validations", JsonUtils.getJsonNodeFromStringWithFallback(
+                    JsonUtils.writeValueAsString(trace.guardrailsValidations())));
+        }
+
+        if (trace.metadata() != null) {
+            enrichedData.put("trace_metadata", trace.metadata());
+        }
+
+        if (trace.usage() != null && !trace.usage().isEmpty()) {
+            enrichedData.put("trace_usage", JsonUtils.getJsonNodeFromStringWithFallback(
+                    JsonUtils.writeValueAsString(trace.usage())));
+        }
+
+        if (trace.totalEstimatedCost() != null) {
+            enrichedData.put("trace_total_estimated_cost", JsonUtils.getJsonNodeFromStringWithFallback(
+                    JsonUtils.writeValueAsString(trace.totalEstimatedCost())));
+        }
+
+        // spanCount and llmSpanCount are primitive int, so they can't be null
+        enrichedData.put("trace_span_count", JsonUtils.getJsonNodeFromStringWithFallback(
+                JsonUtils.writeValueAsString(trace.spanCount())));
+        enrichedData.put("trace_llm_span_count", JsonUtils.getJsonNodeFromStringWithFallback(
+                JsonUtils.writeValueAsString(trace.llmSpanCount())));
+
+        if (trace.duration() != null) {
+            enrichedData.put("trace_duration", JsonUtils.getJsonNodeFromStringWithFallback(
+                    JsonUtils.writeValueAsString(trace.duration())));
+        }
+
+        // Return a new DatasetItem with enriched data
+        return item.toBuilder().data(enrichedData).build();
```

Comment:
Here, we have two options: either update the reference or avoid updating the list we are iterating, and then return a new list.